### PR TITLE
[ci] Add short failing test summary for pytests

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

It is sometimes hard to find all failing tests in buildkite output logs - even filtering for "FAILED" is cumbersome as the output can be overloaded. This PR adds a small utility to add a short summary log in a separate output section at the end of the buildkite job.

The only shared directory between the Buildkite host machine and the test docker container is `/tmp/artifacts:/artifact-mount`. Thus, we write the summary file to this directory, and delete it before actually uploading it as an artifact in the `post-commands` hook.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .buildkite/hooks/post-command [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# In contrast, our build jobs are run in Docker containers.
+# This means that even though our build jobs write to
+# `/artifact-mount`, the directory on the host machine is
+# actually `/tmp/artifacts`.
+# Here, we cat all text files in artifact-mount/test-summaries.
+# These files contain a condensed summary of all failing pytest
+# tests to make it easy for users to see which tests are failing.
+# Because we print them here, we don't need to persist them as
+# artifacts, so we remove the directory afterwards.
+set -e
```
Comment (krfricke) [inline .buildkite/hooks/post-command:12]: Let's only use `set -e` here, please (not `set -x`), as this is about readable output, and `set -x` would defeat that purpose

File: python/ray/tune/syncer.py [hunk 2]
```diff
@@ -98,13 +97,12 @@ def validate_sync_config(sync_config: "SyncConfig"):
         sync_config.node_sync_period = -1
         sync_config.cloud_sync_period = -1
 
-        warnings.warn(
+        raise DeprecationWarning(
```
Comment (krfricke) [inline python/ray/tune/syncer.py:100]: These have been added to trigger a failure for demonstration purposes.

These changes will also be introduced in https://github.com/ray-project/ray/pull/24076 so after https://github.com/ray-project/ray/pull/24076 is merged these changes will disappear and we can just merge.

**Comment by krfricke** [comment]
See here: https://buildkite.com/ray-project/ray-builders-pr/builds/30125#54d81bed-7f76-4ada-8899-3d2ca364e063

**Comment by simon-mo** [comment]
Can we look into buildkite annotation instead of just cat them out? It will improve the UX even more!

**Comment by krfricke** [comment]
Pretty happy with the output @simon-mo https://buildkite.com/ray-project/ray-builders-pr/builds/30291#f76ebe38-f65a-4088-93d5-12fbe3e0b03c
lmk what you think.

Currently only works for pytest-based outputs, so e.g. not example scripts

File: .buildkite/hooks/post-command (hunk 1 above)
Comment (simon-mo) [reply .buildkite/hooks/post-command:12]: can you leave a comment in the script? so we can prevent adding this in the future

**Review by simon-mo — APPROVED** [review]
please add the comment in post-command!

**Comment by simon-mo** [comment]
oh also i think we should have code owners for the hooks file, can you add them in a separate PR?

**Comment by krfricke** [comment]
See here: https://github.com/ray-project/ray/pull/24187

**Pull request merged**